### DISCUSSION TITLE
MGIS for conda on OSX (Apple silicon)

### DISCUSSION THREAD

**Discussion by rutwikg** [thelfer/MFrontGenericInterfaceSupport#139] [Q&A]
Hello everyone,

I have been struggling to get MGIS work with conda environments on macos. I successfully installed the libraries and can import them outside conda environments but cannot do the same with any of my already setup envs. Could someone kindly guide me here.

Thanks and regards
Rutwik

**Comment by thelfer** [comment]
Dear @rutwikg,
Thanks for your interest for `MGIS`. Can you provide details on how to reproduce the issue ?
Best,
Thomas

**Reply by rutwikg** [reply]
Dear @thelfer ,

Thank you so much for the quick response. 

I already use my M3 mac for AI-based projects and wanted to get into mgis and fenics. I installed fenics using conda since it was the easiest way. I built mgis using cmake and exported it to path, also the sitepackages. I can successfully import mgis in python (without conda). But I fail to import mgis in any of the conda envs. I even tried importing it in the python venv that I use for AI and still failed to do so. Since there are no mgis conda packages for osx , I don't know how else I can make it work with virtual envs. 

Hope this is what you need.

Regards
Rutwik

**Reply by thelfer** [reply]
I'll give it try tomorrow.

**Reply by rutwikg** [reply]
Thank you. I will wait for your message.

Have a nice evening.

Regards
Rutwik

**Reply by thelfer** [reply]
>  I built mgis using cmake and exported it to path,

Did you compile `mgis` with `conda`'s `python` ? Are you sure that the versions match ?

**Reply by thelfer** [reply]
I tried on `linux` on a fresh install of `miniconda`. Things are a bit tricky due to `conda`'s oddities, but it seems to work in the end. Here are the steps that I followed:

~~~bash
$ conda install -c conda-forge cxx-compiler
$ conda install conda-forge::boost
$ git clone https://github.com/thelfer/MFrontGenericInterfaceSupport
$ mkdir build
$ cd build
$ cmake ../MFrontGenericInterfaceSupport/ -DCMAKE_BUILD_TYPE=Release -Denable-python-bindings=ON -DPYTHON_INCLUDE_DIR=/home/th202608/codes/miniconda3/install/include/python3.12 -DPYTHON_LIBRARY=/home/th202608/codes/miniconda3/install/lib/libpython3.12.so -DUSE_EXTERNAL_COMPILER_FLAGS=ON -DCMAKE_INSTALL_PREFIX=/home/th202608/codes/miniconda3/tests/mgis/install
$ make -j 8
$ make install
~~~

Would you try to reproduce this on `Mac OS` ? Of course, you have to adapt to update the various paths.

**Reply by rutwikg** [reply]
Hello @thelfer 

I have tried the above steps. still couldn't import the module in conda env.

cmake ../MFrontGenericInterfaceSupport -DCMAKE_BUILD_TYPE=Release -Denable-python-binding=ON -DPYTHON_INCLUDE_DIR=/opt/miniconda3/include/python3.12 -DPYTHON_LIBRARY=/opt/miniconda3/lib/libpython3.12.dylib -DUSE_EXTERNAL_COMPILER_FLAGS=ON -DCMAKE_INSTALL_PREFIX=/opt/miniconda3/tests/mgis/install

I could find the mgis and _mgis.so in site packages of python which I added to path but couldn't find something similar in conda python site packages.

Could this be a problem? 

Regards
Rutwik

**Reply by thelfer** [reply]
I don't know. I updated my environment as follows (in `conda`'s env):

~~~bash
export MGISHOME=/home/th202608/codes/miniconda3/tests/mgis/install
export PATH=$MGISHOME/bin:$PATH
export LD_LIBRARY_PATH=$MGISHOME/lib:$LD_LIBRARY_PATH
export PYTHONPATH=$MGISHOME/lib/python3.12/site-packages/:$PYTHONPATH
~~~

Then `import mgis` worked

**Reply by rutwikg** [reply]
Hello Thomas,

Tried everything. Still doesn't work. I could get it to work in wsl last week with Conda. Had missed some exports but a quick google search took me to your past threads and could make it work in the end. May be it has something to do with OSX. I even tried the homebrew installation of mgis. That didn't help too.

Regards
Rutwik

**Reply by thelfer** [reply]
I have a `Mac Os` box at work. I'll give a try tomorow.

**Reply by rutwikg** [reply]
Thank you so much. That would be a great help. Thank you for getting back so quickly. 

Have a nice evening.

Regards
Rutwik

**Reply by thelfer** [reply]
Hi @rutwikg,
Unfortunately, I did not log on the Mac OS box for quite a long time and lost the cached credentials and they are not automatically reset as they should. My administrators will have to investigate, but this won't be before next week.

**Reply by rutwikg** [reply]
Hello @thelfer 

Thank you for your effort to help me solve my issue. I will gladly wait for your answer and meantime try solving it on my own as well.

Thank you again.

Regards
Rutwik

**Reply by maxence-wz** [reply]
Hello @rutwikg,

I tested this on an Intel MacBook running macOS 10.15.7. Since my Mac is a bit old, I had to install some additional prerequisites, which may not be necessary for your M3 Mac. You can try the following steps and skip any that don’t seem relevant to your setup.

After installing and sourcing Miniconda, I installed these prerequisites:

```{bash}
conda install -c conda-forge cxx-compiler
conda install --solver=classic conda-forge::conda-libmamba-solver conda-forge::libmamba conda-forge::libmambapy conda-forge::libarchive
conda config --set solver classic
conda install conda-forge::boost
```

To install MGIS, follow these steps (ensure you modify the directories to match your system's configuration):
```{bash}
git clone https://github.com/thelfer/MFrontGenericInterfaceSupport
mkdir build
cd build
cmake ../MFrontGenericInterfaceSupport/ -DCMAKE_BUILD_TYPE=Release -Denable-python-bindings=ON -DPYTHON_INCLUDE_DIR=/tmp/miniconda3/include/python3.12 -DPYTHON_LIBRARY=/tmp/miniconda3/lib/libpython3.12.dylib -DUSE_EXTERNAL_COMPILER_FLAGS=ON -DCMAKE_INSTALL_PREFIX=/tmp/MGIS/miniconda3/install
make -j 8
make install
```

Next, update your Conda environment (again, make sure to adjust the paths according to your setup):
```{bash}
export MGISHOME=/tmp/MGIS/miniconda3/install
export PATH=$MGISHOME/bin:$PATH
export LD_LIBRARY_PATH=$MGISHOME/lib:$LD_LIBRARY_PATH
export PYTHONPATH=$MGISHOME/lib/python3.12/site-packages/:$PYTHONPATH
```

Finally, it should work when you run Python and test the installation with:
```{python}
import mgis
```
